kratos/utilities/dof_utilities/dof_array_utilities.h

```diff
@@ -0,0 +1,151 @@
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Ruben Zorrilla
+//                   Riccardo Rossi
+//
+
+#pragma once
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "includes/model_part.h"
+
+namespace Kratos
+{
+
+///@addtogroup KratosCore
+///@{
+
+///@name Kratos Classes
+///@{
+
+/**
+ * @class DofArrayUtilities
+ * @ingroup KratosCore
+ * @brief Utility class for handling the build of DOFs
+ * @details This static class collects some functions to handle the DOFs.
+ * @author Ruben Zorrilla
+ */
+class KRATOS_API(KRATOS_CORE) DofArrayUtilities
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    /// Pointer definition of DofArrayUtilities
+    KRATOS_CLASS_POINTER_DEFINITION(DofArrayUtilities);
+
+    /// Definition of the index type
+    using IndexType = std::size_t;
+
+    /// DOFs array type definition from ModelPart (note that this is PointerVectorSet<DofType>)
+    using DofsArrayType = ModelPart::DofsArrayType;
+
+    /// DOFs vector type definition from ModelPart (note that this is std::vector<DofType::Pointer>)s
+    using DofsVectorType = ModelPart::DofsVectorType;
+
+    /// Auxilary DOF set type definition
+    using AuxiliaryDofsSetType = std::unordered_set<Node::DofType::Pointer, DofPointerHasher>;
+
+    /// Auxiliary slave to master(s) map for slave DOFs
+    using SlaveToMasterDofsMap = std::unordered_map<typename Node::DofType::Pointer, DofsVectorType>;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor.
+    DofArrayUtilities() = delete;
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * @brief Set the Up Dof Array object
+     * This function fills a sorted DOFs array with the DOFs of the provided model part (elements, conditions and constraints)
+     * @param rModelPart The model part containing the entities from which the DOFs are to be taken
+     * @param rDofArray The DOFs array to be filled
+     * @param EchoLevel The echo level (i.e., verbosity level)
+     */
+    static void SetUpDofArray(
+        const ModelPart& rModelPart,
+        DofsArrayType& rDofArray,
+        const unsigned int EchoLevel = 0);
+
+    /**
+     * @brief Set the Up Effective Dof Array and Slave To Master Dofs objects
+     * This function fills a map relating each slave DOF with a vector containing its master DOF(s)
+     * If there are constraints, the effective (i.e., non-slave) DOF array is set based on this map
+     * If there are no constraints, the effective DOF array is the standard DOF array
+     * @param rModelPart The model part containing the constraints
+     * @param rDofArray The already filled and sorted DOF array
+     * @param rEffectiveDofArray The effective DOF array to be filled
+     * @param rSlaveToMasterDofsMap The slave to master DOFs map to be filled
+     * @param EchoLevel The echo level (i.e., verbosity level)
+     */
```

Comment:
Small question, i am little bit confused still about the effective dofs array. Could you put an example in the docs?

Reply:
In short, an effective DOF is any non-slave DOF. Hence, the effective DOFs array is the PVS containing these. Once the final design is merged, I'll properly document it using a toy example (indeed, some of the upcoming tests can be used for this).